File: app/src/main/java/com/jcjm/desafioandroid/viewmodel/PullRequestViewModel.java
package com.jcjm.desafioandroid.viewmodel;

import android.arch.lifecycle.MutableLiveData;
import android.arch.lifecycle.ViewModel;

import com.jcjm.desafioandroid.repository.Repository;
import com.jcjm.desafioandroid.util.ApiResponse;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.schedulers.Schedulers;

/**
 * Created by ${Saquib} on 03-05-2018.
 */

public class PullRequestViewModel extends ViewModel {

    private Repository repository;
    private final CompositeDisposable disposables = new CompositeDisposable();
    private final MutableLiveData<ApiResponse> responseLiveData = new MutableLiveData<>();


    public PullRequestViewModel(Repository repository) {
        this.repository = repository;
    }

    public MutableLiveData<ApiResponse> pullRequestResponse() {
        return responseLiveData;
    }


    public void hitPullRequestsApi(String login, String p_repository) {

        disposables.add(repository.executePullRequestApi(login, p_repository)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .doOnSubscribe((d) -> responseLiveData.setValue(ApiResponse.loading()))
                .subscribe(
                        result -> responseLiveData.setValue(ApiResponse.success(result)),
                        throwable -> responseLiveData.setValue(ApiResponse.error(throwable))
                ));

    }

    @Override
    protected void onCleared() {
        disposables.clear();
    }
}

File: app/src/main/java/com/concrete/andresdavid/desafioandroid/viewmodel/PullRequestViewModel.kt
package com.concrete.andresdavid.desafioandroid.viewmodel

import android.arch.lifecycle.LiveData
import android.arch.lifecycle.MutableLiveData
import android.arch.lifecycle.ViewModel
import com.concrete.andresdavid.desafioandroid.model.PullRequest
import com.concrete.andresdavid.desafioandroid.model.Resource
import com.concrete.andresdavid.desafioandroid.repository.PullRequestRepository
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers


class PullRequestViewModel : ViewModel() {
    private val pullRequestRepo: PullRequestRepository= PullRequestRepository()
    private lateinit var pullRequestsData: MutableLiveData<Resource<List<PullRequest>>>

    fun getPullRequests(repoName: String): LiveData<Resource<List<PullRequest>>> {
        if (!::pullRequestsData.isInitialized) {
            pullRequestsData = MutableLiveData()
            load(repoName)
        }
        return pullRequestsData
    }

    fun load(repoName: String) {
        pullRequestRepo.getPullRequestsByRepository(repoName)
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeOn(Schedulers.io())
                .subscribe({ result ->
                    addItems(result)
                }, { error ->
                    this.pullRequestsData.value = Resource.error(error.message!!, mutableListOf())
                })
    }

    fun addItems(newItems: List<PullRequest>){
        this.pullRequestsData.value = Resource.success(newItems)
    }

}